espnet/asr/pytorch_backend/asr.py

```diff
@@ -59,6 +61,87 @@
 REPORT_INTERVAL = 100
 
 
+def time_warp(spec, W=5):
+    """Time warping
+
+    :param spec: input tensor with shape (T, dim)
+    """
+    spec = spec.unsqueeze(0)
+    num_rows = spec.shape[1]
+    spec_len = spec.shape[2]
+    device = spec.device
+
+    y = num_rows // 2
+    horizontal_line_at_ctr = spec[0][y]
+    assert len(horizontal_line_at_ctr) == spec_len
+
+    point_to_warp = horizontal_line_at_ctr[random.randrange(W, spec_len - W)]
+    assert isinstance(point_to_warp, torch.Tensor)
+
+    # Uniform distribution from (0,W) with chance to be up to W negative
+    dist_to_warp = random.randrange(-W, W)
+    src_pts, dest_pts = (torch.tensor([[[y, point_to_warp]]], device=device),
+                         torch.tensor([[[y, point_to_warp + dist_to_warp]]], device=device))
+    warped_spectro, dense_flows = sparse_image_warp(spec, src_pts, dest_pts)
+    return warped_spectro.squeeze(3).squeeze(0)
+
+
+def freq_mask(spec, F=30, num_masks=1, replace_with_zero=False):
+    """Frequency masking
+
+    :param spec: input tensor with shape (T, dim)
+    """
+    cloned = spec.unsqueeze(0).clone()
+    num_mel_channels = cloned.shape[1]
+
+    for i in range(0, num_masks):
+        f = random.randrange(0, F)
+        f_zero = random.randrange(0, num_mel_channels - f)
+
+        # avoids randrange error if values are equal and range is empty
+        if (f_zero == f_zero + f):
+            return cloned.squeeze(0)
+
+        mask_end = random.randrange(f_zero, f_zero + f)
+        if (replace_with_zero):
+            cloned[0][f_zero:mask_end] = 0
+        else:
+            cloned[0][f_zero:mask_end] = cloned.mean()
+    return cloned.squeeze(0)
+
+
+def time_mask(spec, T=40, num_masks=1, replace_with_zero=False):
+    """Time masking
+
+    :param spec: input tensor with shape (T, dim)
+    """
+    cloned = spec.unsqueeze(0).clone()
+    len_spectro = cloned.shape[2]
+
+    for i in range(0, num_masks):
+        t = random.randrange(0, T)
+        t_zero = random.randrange(0, len_spectro - t)
+
+        # avoids randrange error if values are equal and range is empty
+        if (t_zero == t_zero + t):
+            return cloned.squeeze(0)
+
+        mask_end = random.randrange(t_zero, t_zero + t)
+        if (replace_with_zero):
+            cloned[0][:, t_zero:mask_end] = 0
+        else:
+            cloned[0][:, t_zero:mask_end] = cloned.mean()
+    return cloned.squeeze(0)
+
+
+def specaug(spec):
+    """SpecAug implementation from https://github.com/zcaceres/spec_augment
```

Comment:
Move these functions to a new individual module and add https://github.com/zcaceres/spec_augment/blob/master/LICENSE to the module-level docstring